cassandra/src/main/java/org/hibernate/ogm/datastore/cassandra/CassandraDialect.java

```diff
@@ -183,14 +200,23 @@ public void insertOrUpdateTuple(EntityKey key, Tuple tuple, TupleContext tupleCo
 
 			// insert and update are both 'upsert' in cassandra.
 			Insert insert = insertInto( quote( key.getTable() ) );
-			Object[] columnValues = new Object[updateOps.size()];
+			List<Object> columnValues = new LinkedList<>();
+			Set<String> seenColNames = new HashSet<>();
 			for ( int i = 0; i < updateOps.size(); i++ ) {
 				TupleOperation op = updateOps.get( i );
 				insert.value( quote( op.getColumn() ), QueryBuilder.bindMarker() );
-				columnValues[i] = op.getValue();
+				columnValues.add( op.getValue() );
+				seenColNames.add( op.getColumn() );
+			}
+			for ( int j = 0; j < key.getColumnNames().length; j++ ) {
```

Comment:
hmm. Those are a) what I'd expect and b) valid, since they work fine from the cqlsh shell. I can cause different errors by dropping the index or the 'allow filtering' clause, but can't reproduce the same failure you're seeing. What version of cassandra are you running against?